File: ruby/component/inputhistory.rb
# Created by Satoshi Nakagawa.
# You can redistribute it and/or modify it under the Ruby's license or the GPL2.

class InputHistory
  
  MAX_SIZE = 50
  
  def initialize
    @buf = []
    @pos = 0
  end
  
  def add(s)
    @pos = @buf.size
    return if s.empty?
    return if @buf[-1] == s
    @buf << s
    @buf.shift if @buf.size > MAX_SIZE
    @pos = @buf.size
  end
  
  def up(s)
    if s && !s.empty?
      cur = @buf[@pos]
      if !cur || cur != s
        # if the text was modified, append the current text
        @buf << s
        if @buf.size > MAX_SIZE
          @buf.shift
          @pos -= 1
        end
      end
    end
    @pos -= 1
    if @pos < 0
      @pos = 0
      nil
    else
      @buf[@pos] || ''
    end
  end
  
  def down(s)
    if !s || s.empty?
      @pos = @buf.size
      return nil
    end
    cur = @buf[@pos]
    if !cur || cur != s
      add(s)
      ''
    else
      @pos += 1
      @buf[@pos] || ''
    end
  end
end

File: Classes/Library/InputHistory.m
// LimeChat is copyrighted free software by Satoshi Nakagawa <psychs AT limechat DOT net>.
// You can redistribute it and/or modify it under the terms of the GPL version 2 (see the file GPL.txt).

#import "InputHistory.h"


#define INPUT_HISTORY_MAX   50


@implementation InputHistory
{
    NSMutableArray* _buf;
    int _pos;
}

- (id)init
{
    self = [super init];
    if (self) {
        _buf = [NSMutableArray new];
    }
    return self;
}

- (void)add:(NSString*)s
{
    _pos = _buf.count;
    if (s.length == 0) return;
    if ([[_buf lastObject] isEqualToString:s]) return;

    [_buf addObject:s];

    if (_buf.count > INPUT_HISTORY_MAX) {
        [_buf removeObjectAtIndex:0];
    }
    _pos = _buf.count;
}

- (NSString*)up:(NSString*)s
{
    if (s && s.length > 0) {
        NSString* cur = nil;
        if (0 <= _pos && _pos < _buf.count) {
            cur = [_buf objectAtIndex:_pos];
        }

        if (!cur || ![cur isEqualToString:s]) {
            // if the text was modified, add it
            [_buf addObject:s];
            if (_buf.count > INPUT_HISTORY_MAX) {
                [_buf removeObjectAtIndex:0];
                --_pos;
            }
        }
    }

    --_pos;
    if (_pos < 0) {
        _pos = 0;
        return nil;
    }
    else if (0 <= _pos && _pos < _buf.count) {
        return [_buf objectAtIndex:_pos];
    }
    else {
        return @"";
    }
}

- (NSString*)down:(NSString*)s
{
    if (!s || s.length == 0) {
        _pos = _buf.count;
        return nil;
    }

    NSString* cur = nil;
    if (0 <= _pos && _pos < _buf.count) {
        cur = [_buf objectAtIndex:_pos];
    }

    if (!cur || ![cur isEqualToString:s]) {
        // if the text was modified, add it
        [self add:s];
        return @"";
    }
    else {
        ++_pos;
        if (0 <= _pos && _pos < _buf.count) {
            return [_buf objectAtIndex:_pos];
        }
        return @"";
    }
}

@end
